src/transformStateWithClones.js

```diff
@@ -8,6 +8,37 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  let currentState = { ...state };
+  const states = [];
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+      case 'addProperties':
+        if (typeof action.extraData === 'object' && action.extraData !== null) {
+          Object.entries(action.extraData).forEach(([key, value]) => {
+            newState[key] = value;
+          });
+        }
+        break;
+
+      case 'removeProperties':
+        if (action.keysToRemove) {
+          for (const key of action.keysToRemove) {
+            delete newState[key];
+          }
+        }
+        break;
+    }
```

Comment:
[DONT REPEAT YOURSELF] - You are pushing `newState` to `states` array in every case of the switch. This action should be done once after the switch statement to avoid repetition.